app/services/amalgamator/processor/separate.rb

```diff
@@ -0,0 +1,117 @@
+# frozen_string_literal: true
+
+module Amalgamator
+  # Subtracts a dataset from the target dataset to create a new dataset
+  class Processor::Separate < Processor::Base
+    class << self
+      # The values for the given datasets are processed in 3 steps:
+      # 1. Subtract the values according to the combination_method set in the InterfaceItem
+      # 2. Round all 'subtracted' values to 8 decimals
+      # 3. Adjust flexible shares if necessary
+      def perform(target_dataset, subtracted_dataset)
+        dataset_c_values = calculate_item_values(target_dataset, subtracted_dataset) # dataset C is the result of subtracting A from B
+        round_item_values(dataset_c_values)
+        dataset_c_values = adjust_flexible_shares(dataset_c_values)
+
+        dataset_c_values
+      end
+
+      # Creates a hash with the item's keys as keys and the values for dataset C as values, e.g.:
+      # {
+      #   number_of_cars: 3,          # C = B - A
+      #   co2_emission_1990: 500,     # C = B - A
+      #   interconnector_capacity: 50 # C = B - A
+      # }
+      def calculate_item_values(target_dataset, subtracted_dataset)
+        InterfaceElement.items.to_h do |item|
+          dataset_b_value = fetch_valid_value(target_dataset, item)
+          dataset_a_value = fetch_valid_value(subtracted_dataset, item)
+
+          # Use the combination method to determine how C is calculated from B and A
+          value_c = case item.nested_combination_method
+            when 'average'
+              calculate_average(dataset_b_value, dataset_a_value)
+            when 'min', 'max' # Just take the value of B in these cases
+              dataset_b_value
+            when ->(cm) { cm.is_a?(Hash) && cm.key?('weighted_average') }
+              calculate_weighted_average(item, target_dataset, subtracted_dataset, dataset_b_value, dataset_a_value)
+            when 'sum', '', nil # Default to 'difference' (C = B - A)
+              dataset_b_value - dataset_a_value
+            else
+              raise ArgumentError, "Unknown combination method '#{item.nested_combination_method}' for InterfaceItem: #{item.key}"
+            end
+
+          [item.key, value_c]
+        end.compact
+      end
```

Comment:
```suggestion
      def calculate_item_values(target_dataset, subtracted_dataset)
        InterfaceElement.items.to_h do |item|
          [item.key, combined_value_for(target_dataset, subtracted_dataset, item)]
        end.compact
      end

      # Use the combination method to determine how C is calculated from B and A
      def combined_value_for(target_dataset, subtracted_dataset, item)
        case item.nested_combination_method
        when 'average'
          calculate_average(
            value_for(target_dataset, item),
            value_for(subtracted_dataset, item)
          )
        when 'min', 'max' # Just take the value of B in these cases
          value_for(target_dataset, item)
        when ->(cm) { cm.is_a?(Hash) && cm.key?('weighted_average') }
          calculate_weighted_average(
            item,
            target_dataset,
            subtracted_dataset,
            value_for(target_dataset, item),
            value_for(subtracted_dataset, item)
          )
        when 'sum', '', nil # Default to 'difference' (C = B - A)
          value_for(target_dataset, item) - value_for(subtracted_dataset, item)
        else
          raise ArgumentError, unknown_method_message_for(item)
        end
      end
```
Nice combination of cases together! We can try to make the lines a bit less long like this. I'm also wondering if we are missing a wrapper object around the datasets to define all these helper methods on, instead of having this one processor "god". But that's not in scope right now!